### ISSUE TITLE
Support for enterprise time zones

### ISSUE THREAD

**Issue by andrewpbrett** [openfoodfoundation/openfoodnetwork#6108]
## What is the problem we are solving 
Some instances have enterprises across multiple time zones (US, Canada, Australia). Currently any time an enterprise does anything with time it's assumed that they're working in the time zone of the server. Many users might not be aware of this and incorrectly set the open/close times for order cycles, which could then be up to a few hours off from what they intended. 

See https://community.openfoodnetwork.org/t/allow-enterprises-to-set-their-own-time-zone/2058

and

https://openfoodnetwork.slack.com/archives/CG7NJ966B/p1600561652011600

## Success factors = expected outcome 
Less annoyance from users who know about the issue. 
Less bug reports from people saying their shop closed early if they don't know about the issue.

## Useful information for inception
Draft PR: https://github.com/openfoodfoundation/openfoodnetwork/pull/6095

This PR the minimum viable feature version of a solution to the problem. We could consider going further by:

Noting when a user's browser is using a different time zone than the enterprise(s) they are managing and directing them to the settings page to change it
Improving the Order Cycle datetime picker to make it more obvious that the open/close times are in a specific timezone at all (I've created dozens of order cycles and never noticed the timezone offset at the end of that string, even though it's very obvious now)
Add support to other object (besides Order Cycle) that rely on times: Order.completed_at, Product.available_on, Shipment.shipped_at, Subscription.begins_at, Subscription.ends_at, Subscription.canceled_at, Subscription.paused_at

In addition, from @luisramos0: 

We need to investigate potential problems this may cause, off the top of my head I can think of three:

- order_cycle dates will impact subscriptions
- order dates will impact reports
- order_cycle dates will have to be read by the shopfront with timezone so that the shop closes at the correct time, right?

**Comment by luisramos0** [comment]
Thanks for creating the issue :+1: 
One more question I have: how is this going to work with users with permissions to see enterprises in multiple timezones? I believe timezone is a display attribute so all users will see all dates in their timezone even if the enterprise is configured to a specific timezone?

**Comment by andrewpbrett** [comment]
That's a great question... it looks like all of these objects tie back to an enterprise. In the PR that's the model I added the `timezone` field to; so if a superadmin is looking at order cycles, they would see something like:

Name | Open | Close
:----   | :----- | :-----
Fredo's West Coast Farm OC | 2020-10-01 23:00 -0500 | 2020-10-20 23:00 -0500
Mary's East Coast Farm OC | 2020-10-01 23:00 -0800 | 2020-10-20 23:00 -0800

And if they changed something, it would be assumed that they were entering the time in the timezone of that enterprise, and not necessarily the time zone they were in.

**Comment by luisramos0** [comment]
Instead of making timezone a property of an enterprise, we could use the timezone of the browser to display times?
https://stackoverflow.com/a/37512371
what do you think about this approach?

**Comment by luisramos0** [comment]
Another alternative is to fetch timezone in JS and set it in the user session just like we do for locale, this way we could make the conversion to the user timezone on the server.

**Comment by andrewpbrett** [comment]
We could detect the browser timezone, yes, but long term I think we'll want to have it be a field of enterprise anyway. That way things like reports and invoices that are generated outside of the browser will be correct for the shop. 

The last time I worked with detecting the browser's timezone it was not 100% reliable; it looks like it has gotten much better since then though: https://caniuse.com/?search=DateTimeFormat. I think we'd still want to have a workaround in case there are edge cases (someone is running some browser extension that messes with detecting the time zone etc.)

Can you tell I don't trust browsers? 😆 

But yes I agree - something on the frontend would be good to have, but I think we should probably have both. #beltandsuspenders

**Comment by luisramos0** [comment]
I am not sure about that argument for still having the timestamp in the enterprise. If the user timezone we get in the browser is not good enough we can default to the instance timezone on those cases. Maybe there are other good reasons to have the timezone on the enterprise! For example, I am particularly curious about how big this piece of work is in total. If it's having timezone on enterprise is a faster solution, I'd accept that as a good reason to go for it.

I'd recommend you hold the implementation until we agree on the approach for this.

Maybe we need to go through the pages and see what needs to be done, in a way we can do Acceptance Tests Driven Development here, let's write the tests before we start the development effort?
There are a few cases but we can consider the most important one: Instance is timezone A and both Enterprise manager and customer are in timezone B.

ping @Matt-Yorkley @sauloperez @mkllnk thoughts?

**Comment by andrewpbrett** [comment]
Agreed on holding off on any further work. One other option would be to consider merging it as it stands now (after testing obviously). The test suite is passing, and it solves the main pain point of order cycles open/close times being incorrect. If we take a look at the three items @luisramos0 mentioned (really just two: reports and subscriptions) and decide that they won't be negatively impacted by this, it could be the quick solution.

**Comment by luisramos0** [comment]
ah, ok, if you think your PR will work by itself I am ok with moving it forward :+1: I thought it would need more work to make it coherent.
We need to verify that the shopfront and subscriptions will see the correct time.

**Comment by andrewpbrett** [comment]
Yes sorry I meant to move it out of draft sooner. I'll check on the other places and see if we show the correct time there. 

I know this one did not follow the process from the presentation this morning, sorry about that! 😆
